### PR TITLE
[R] Use `type` argument to control prediction types

### PR DESCRIPTION
**Pull request by david-cortes**
ref #7906 

In R packages for statistical modeling, the `predict` functions for model objects tend to allow switching between different prediction types such as probabilities or classes through a parameter `type` (for example, base R's `predict.glm` does it like that).

`xgboost` offers several mutually exclusive parameters to control prediction types (`outputmargin`, `precontrib`, `predinteraction`, `predleaf`). This PR merges them all into a single `type` parameter in line with base R and other popular packages, adding an option to predict classes for classification objectives.

Note however that objective `multi:softmax` is problematic in this scenario, since it doesn't allow switching between probabilities and classes on the fly, so an exception about it is mentioned in the docs.

Right now the code logic is to check the objective and determine if and how it should determine the predicted class from the outputs, but:
* I wasn't able to find any C-level function to extract the booster's objective, and am not sure that the way is done in this PR is ideal.
* It currently lists a set of objectives which allow class prediction from the original output, but I think a better approach would be to check if the objective starts with `binary` or `multi` - problem with such approach right now is that `multi:softmax` doesn't output probabilities.

Disclaimer: this is likely to cause several breakages in R packages that depend on `xgboost`.

### CODE REVIEW

**Comment by trivialfis** [comment]
Thank you for the work on aligning XGBoost with the R ecosystem. I will take a deeper look later to see if it's possible to keep the compatibility with some deprecation warnings. Otherwise we will have to run a reverse dependency check and talk to other projects before cran submission for the next release.

**Comment by trivialfis** [comment]
@hetong007 Would you great if you can help review the changes.

**Comment by hetong007** [comment]
In terms of the potential breaks, may I suggest that (at least in the next release) we keep both the new argument `type` and the old set of arguments? In this case, users can work with `predict` with either way. Just when users predict with the old arguments, `predict` throws out a deprecation warning.

Moreover, by setting a higher priority to `type` we resolve the case where the new and the old are conflicting.

**Comment by david-cortes** [comment]
Added deprecation warnings and acceptance of the old argument types.

File: R-package/R/xgb.Booster.R [hunk 1]
```diff
@@ -375,23 +395,78 @@ predict.xgb.Booster <- function(object, newdata, missing = NA, outputmargin = FA
...
+    old_args_for_type <- list(
+      "outputmargin" = "margin",
+      "predleaf" = "leaf",
+      "predcontrib" = "contrib",
+      "predinteraction" = "interaction"
+    )
+
+    for (arg in names(old_args_for_type)) {
+      if (arg %in% extra_argnames) {
+        warning(
```
Comment (hetong007) [inline R-package/R/xgb.Booster.R:412]: My only concern with `warning` is that it might break CRAN checks of reverse dependencies. We need to talk to their maintainers before pushing this change to CRAN.

Comment (david-cortes) [reply R-package/R/xgb.Booster.R:412]: It most definitly will. Good thing is, next version is going to be a 2.0, which signals a major release and thus potentially breaking changes.

Comment (hetong007) [reply R-package/R/xgb.Booster.R:412]: In concept for general development this is true, while for CRAN it still regardlessly breaks reverse dependencies and got Ripleyed.

Comment (mayer79) [reply R-package/R/xgb.Booster.R:412]: Changing the interface of `predict()` will definitively cause chaos, not only on CRAN.

**Comment by david-cortes** [comment]
Renamed the types to be more in line with base R.
